File: rxbinding-recyclerview-v7/src/main/java/com/jakewharton/rxbinding2/support/v7/widget/RecyclerViewChildAttachStateChangeEventObservable.java
package com.jakewharton.rxbinding2.support.v7.widget;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.OnChildAttachStateChangeListener;
import android.view.View;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class RecyclerViewChildAttachStateChangeEventObservable extends Observable<RecyclerViewChildAttachStateChangeEvent> {
  private final RecyclerView view;

  RecyclerViewChildAttachStateChangeEventObservable(RecyclerView recyclerView) {
    this.view = recyclerView;
  }

  @Override protected void subscribeActual(Observer<? super RecyclerViewChildAttachStateChangeEvent> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnChildAttachStateChangeListener(listener);
  }

  final class Listener extends MainThreadDisposable implements OnChildAttachStateChangeListener {
    private final RecyclerView recyclerView;
    private final Observer<? super RecyclerViewChildAttachStateChangeEvent> observer;

    Listener(RecyclerView recyclerView, Observer<? super RecyclerViewChildAttachStateChangeEvent> observer) {
      this.recyclerView = recyclerView;
      this.observer = observer;
    }

    @Override public void onChildViewAttachedToWindow(View childView) {
      if (!isDisposed()) {
        observer.onNext(RecyclerViewChildAttachEvent.create(recyclerView, childView));
      }
    }

    @Override public void onChildViewDetachedFromWindow(View childView) {
      if (!isDisposed()) {
        observer.onNext(RecyclerViewChildDetachEvent.create(recyclerView, childView));
      }
    }

    @Override protected void onDispose() {
      recyclerView.removeOnChildAttachStateChangeListener(this);
    }
  }
}

File: rxbinding-recyclerview/src/main/java/com/jakewharton/rxbinding3/recyclerview/RecyclerViewChildAttachStateChangeEventObservable.kt
@file:JvmMultifileClass
@file:JvmName("RxRecyclerView")

package com.jakewharton.rxbinding3.recyclerview

import android.view.View
import androidx.annotation.CheckResult
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnChildAttachStateChangeListener
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding2.internal.checkMainThread

/**
 * Create an observable of child attach state change events on `recyclerView`.
 *
 * *Warning:* The created observable keeps a strong reference to `recyclerView`.
 * Unsubscribe to free this reference.
 */
@CheckResult
fun RecyclerView.childAttachStateChangeEvents(): Observable<RecyclerViewChildAttachStateChangeEvent> =
    RecyclerViewChildAttachStateChangeEventObservable(this)

private class RecyclerViewChildAttachStateChangeEventObservable(
  private val view: RecyclerView
) : Observable<RecyclerViewChildAttachStateChangeEvent>() {

  override fun subscribeActual(observer: Observer<in RecyclerViewChildAttachStateChangeEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(
        view, observer)
    observer.onSubscribe(listener)
    view.addOnChildAttachStateChangeListener(listener)
  }

  class Listener(
    private val recyclerView: RecyclerView,
    private val observer: Observer<in RecyclerViewChildAttachStateChangeEvent>
  ) : MainThreadDisposable(), OnChildAttachStateChangeListener {

    override fun onChildViewAttachedToWindow(childView: View) {
      if (!isDisposed) {
        observer.onNext(RecyclerViewChildAttachEvent(recyclerView, childView))
      }
    }

    override fun onChildViewDetachedFromWindow(childView: View) {
      if (!isDisposed) {
        observer.onNext(RecyclerViewChildDetachEvent(recyclerView, childView))
      }
    }

    override fun onDispose() {
      recyclerView.removeOnChildAttachStateChangeListener(this)
    }
  }
}
